zerver/views/messages.py

```diff
@@ -603,11 +603,33 @@ def get_old_messages_backend(request, user_profile,
 
 @has_request_variables
 def update_message_flags(request, user_profile,
-                      messages=REQ('messages', validator=check_list(check_int)),
-                      operation=REQ('op'), flag=REQ('flag'),
-                      all=REQ('all', validator=check_bool, default=False)):
+                         messages=REQ('messages', validator=check_list(check_int)),
+                         operation=REQ('op'), flag=REQ('flag'),
+                         all=REQ('all', validator=check_bool, default=False),
+                         stream_name=REQ('stream_name', default=None),
+                         topic_name=REQ('topic_name', default=None)):
+
     request._log_data["extra"] = "[%s %s]" % (operation, flag)
-    do_update_message_flags(user_profile, operation, flag, messages, all)
+    stream = None
+    if stream_name is not None:
+        stream = get_stream(stream_name, user_profile.realm)
+        if not stream:
+            raise JsonableError('No such stream \'%s\'' %  (stream_name, ))
+        user_subscribed = Subscription.objects.filter(user_profile=user_profile,
+                                                      recipient__type_id=stream.id,
+                                                      recipient__type=Recipient.STREAM,
+                                                      active=True,
+                                                      ).exists()
+        if not user_subscribed:
+            raise JsonableError('User not subscribed to stream \'%s\', can\'t mark as read' %  (stream_name, ))
```

Comment:
I think this logic has a small bug.  If a user was originally subscribed to a stream, received some messages they hadn't read, and then unsubscribed from the stream, they should still be able to mark those messages as read. 

I think the fix would be to either drop this check or make it work like the topic one just below (where you check if the user ever received any messages on the stream).